### ISSUE TITLE
Native controls cue overlap avoidance works badly for vertical

### ISSUE THREAD

**Issue by zcorpan** [w3c/webvtt#276]
https://w3c.github.io/webvtt/#processing-model

> If the user agent is exposing a user interface for video, add to output one or more completely transparent positioned CSS block boxes that cover the same region as the user interface\.

Most desktop browsers at least have a horizontal controls bar at the bottom of the video, showing and hiding based on hover and paused state. This works fine for horizontal cues. But for vertical cues, assuming the default size:100%, it means they don't fit anywhere.

I suppose it's similar if you have a horizontal cue and a vertical cue active at the same time, which is not unreasonable if you have e.g. English and vertical Japanese tracks both activated.

https://zcorpan.github.io/live-webvtt-viewer/#vtt=WEBVTT%0A%0A00%3A00%3A00.001+--%3E+00%3A00%3A10.000%0AHorizontal+1%0A%0A00%3A00%3A00.000+--%3E+00%3A00%3A10.000+vertical%3Alr+size%3A100%25%0AVertical+1%0A%0A00%3A00%3A01.000+--%3E+00%3A00%3A10.000%0AHorizontal+2

**Comment by zcorpan** [comment]
Ideas for how to fix this?

Should overlap avoidance only happen vertically for horizontal cues, and only horizontally for vertical cues? (It seems bad to always overlap, though, if horizontal and vertical 100% size cues are active at the same time.)

Should cues shrink to make place for already-active cues that are using an orthogonal writing-mode?

**Comment by dwsinger** [comment]
> On Dec 1, 2015, at 6:35 , Simon Pieters <notifications@github.com> wrote:
> 
> Ideas for how to fix this?
> 
> Should overlap avoidance only happen vertically for horizontal cues, and only horizontally for vertical cues? (It seems bad to always overlap, though, if horizontal and vertical 100% size cues are active at the same time.)
> 
> Should cues shrink to make place for already-active cues that are using an orthogonal writing-mode?

Surely if you are expecting horizontal and vertical cues to be simultaneously active, you’d set the vertical cue region’s bottom sufficiently high that the horizontal cues don’t encroach, and/or set the horizontal cue region’s right so that they don’t encroach?


David Singer
Manager, Software Standards, Apple Inc.

**Comment by zcorpan** [comment]
What if they are separate tracks, e.g. one is a translation of the other? If we can assume that different tracks can make space for each other,  we could remove the overlap avoidance altogether. But I think that isn't particularly reasonable. It seems reasonable to be able to have an arbitrary number of available tracks with default settings, and let the user activate any two and have them not overlap each other. From that perspective, it seems reasonable to be able to use vertical text in one of the tracks without having to guess a "safe" size for its cues to make room for other tracks *and* the native controls.

**Comment by silviapfeiffer** [comment]
Overlap between different tracks is a user case that in my experience falls below the 2% usage. We're should do overlap avoidance with controls, because that's 100% use case. We might do overlap avoidance between cues within vtt file because that provides nice features (such as the scrolling, or avoid overlap when users change font size). Everything else is over engineering and should be solved by the author. In fact, html only allows one track to be visible at any time anyway, so you'd have to force the rendering.

**Comment by dwsinger** [comment]
> On Dec 1, 2015, at 14:21 , Silvia Pfeiffer <notifications@github.com> wrote:
> 
> Overlap between different tracks is a user case that in my experience falls below the 2% usage. We're should do overlap avoidance with controls, because that's 100% use case. We might do overlap avoidance between cues within vtt file because that provides nice features (such as the scrolling, or avoid overlap when users change font size). Everything else is over engineering and should be solved by the author. In fact, html only allows one track to be visible at any time anyway, so you'd have to force the rendering.

I agree.  The question of how *compositions* of more than one track work is up to the composing system, not the coding system of any of the tracks (IMHO).

David Singer
Manager, Software Standards, Apple Inc.

**Comment by zcorpan** [comment]
> In fact, html only allows one track to be visible at any time anyway, so you'd have to force the rendering.

What do you mean? HTML certainly supports enabling multiple tracks, both with script and by the user. Automatic track selection will only enable one subtitle/caption, but that seems pretty reasonable.

**Comment by silviapfeiffer** [comment]
> HTML certainly supports enabling multiple tracks, both with script and by the user. Automatic track selection will only enable one subtitle/caption, but that seems pretty reasonable.

Hmmm I guess you're right and it's just for default setting. You can activate multiple through Javascript or user settings. However, browser and video players that have implemented track selection menus typically disable a previously chosen visible track (captions, subtitles) when activating a new one. I'm going for the main use case here.

**Comment by zcorpan** [comment]
I agree that it's not a particularly common use case. But it is a use case that WebVTT and the timed text model in HTML has been designed to support.

![Example with two tracks enabled](http://damowmow.com/temp/mpc-storm-codec.JPG)

https://wiki.whatwg.org/wiki/Use_cases_for_timed_tracks_rendered_over_video_by_the_UA

Should we not make this possible at all? How should it be done?

**Comment by silviapfeiffer** [comment]
It's totally possible now. Just that the author of the WebVTT files has to make sure they author the files so they can be displayed at the same time through explicit positioning.

I guess your suggestion to do overlap avoidance for horizontal cues with other horizontal cues only and similarly vertical with vertical only makes the most sense.

**Comment by dwsinger** [comment]
I'm pretty uncomfortable with the idea that we should step out of the 'sandbox' of considering a single VTT file, and look at the interactions of multiple VTT files (and, ugh, maybe VTT files with other files in the same session.) Can we keep this open and punt it to 'future consideration'?

**Comment by foolip** [comment]
FWIW, I've actually encountered this problem when fiddling with Blink. Hovering the video will cause the controls to show, and they can make the vertical text disappear or move someplace far, far away.
